### PR TITLE
Avoid scheduling when DiscoveryClient.shutdown called

### PR DESCRIPTION
**Pull request by jingege**
#973

### CODE REVIEW

**Comment by qiangdavidliu** [comment]
@jingege thanks for the PR, that makes sense. I believe you still need to call scheduler.shutdown() before blocking on an awaitTermination however. See the last example at the [ExecutorService javadoc](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html). The phased shutdown example in the javadoc seems to be the right way to execute the shutdown, would you mind updating your PR to reflect that?

Appreciate the PR, thanks.

**Comment by jingege** [comment]
@qiangdavidliu Thx for your advice. I updated the PR.

File: eureka-client/src/main/java/com/netflix/discovery/InstanceInfoReplicator.java [hunk 1]
```diff
@@ -68,10 +69,27 @@ public void start(int initialDelayMs) {
     }
 
     public void stop() {
-        scheduler.shutdownNow();
+        Future latestPeriodic = scheduledPeriodicRef.get();
+        if (latestPeriodic != null && !latestPeriodic.isDone()) {
+            logger.info("Canceling the latest scheduled before stop");
+            boolean result = latestPeriodic.cancel(false);
+            System.out.println("Result:" + result);
```
Comment (rs017991) [inline eureka-client/src/main/java/com/netflix/discovery/InstanceInfoReplicator.java]: If keeping this, should use logger and reformat the message to give more context.

File: eureka-client/src/main/java/com/netflix/discovery/InstanceInfoReplicator.java [hunk 2]
```diff
@@ -68,10 +69,27 @@ public void start(int initialDelayMs) {
     }
 
     public void stop() {
-        scheduler.shutdownNow();
+        Future latestPeriodic = scheduledPeriodicRef.get();
+        if (latestPeriodic != null && !latestPeriodic.isDone()) {
+            logger.info("Canceling the latest scheduled before stop");
+            boolean result = latestPeriodic.cancel(false);
```
Comment (qiangdavidliu) [inline eureka-client/src/main/java/com/netflix/discovery/InstanceInfoReplicator.java]: You shouldn't need to cancel the future here, shutting down the scheduler should handle that as the future is scheduled on that.

**Comment by qiangdavidliu** [comment]
@jingege apologies for the late review, I have been away. Some minor comments.

**Comment by abracadv8** [comment]
@qiangdavidliu - This user has not been active since this pull request.  I created a second pull request based on their work #1023 - Can you review that?

**Comment by jingege** [comment]
@qiangdavidliu Very sorry for not responding this issue so long. I was so busy these days and missed the notification. @abracadv8 Thank you very much for finishing this PR.

**Pull request merged**